File: cpp/conformance/command_writer.cpp
#include "conformance/command_writer.hpp"

#include <stdexcept>

#include "io_github_giovanicaprison_matching_protocol/CancelOrder.h"
#include "io_github_giovanicaprison_matching_protocol/InstrumentDefinition.h"
#include "io_github_giovanicaprison_matching_protocol/MassCancel.h"
#include "io_github_giovanicaprison_matching_protocol/MessageHeader.h"
#include "io_github_giovanicaprison_matching_protocol/NewOrder.h"
#include "io_github_giovanicaprison_matching_protocol/ReplaceOrder.h"
#include "io_github_giovanicaprison_matching_protocol/SessionStateChange.h"

namespace io::github::giovanicaprison::matching::conformance {

namespace {

using protocol::AllocationAlgorithm;
using protocol::MessageHeader;
using protocol::PricingInstruction;
using protocol::SessionState;
using protocol::Side;
using protocol::TimeInForce;

constexpr std::uint32_t INSTRUMENT_ID = 1;
constexpr int DEFAULT_PARTICIPANT = 1;
constexpr char ABSENT[] = "-";

std::unordered_map<std::string, std::string> options(const std::vector<std::string>& words,
                                                     const std::size_t from) {
  std::unordered_map<std::string, std::string> found;
  for (std::size_t at = from; at < words.size(); at++) {
    const std::size_t equals = words[at].find('=');
    if (equals != std::string::npos && equals > 0) {
      found.emplace(words[at].substr(0, equals), words[at].substr(equals + 1));
    }
  }
  return found;
}

std::int64_t number(const std::unordered_map<std::string, std::string>& options,
                    const std::string& key) {
  const auto found = options.find(key);
  if (found == options.end()) {
    throw std::runtime_error("missing " + key);
  }
  return std::stoll(found->second);
}

std::int64_t number(const std::unordered_map<std::string, std::string>& options,
                    const std::string& key, const std::int64_t fallback) {
  const auto found = options.find(key);
  return found == options.end() ? fallback : std::stoll(found->second);
}

std::string word(const std::unordered_map<std::string, std::string>& options,
                 const std::string& key) {
  const auto found = options.find(key);
  if (found == options.end()) {
    throw std::runtime_error("missing " + key);
  }
  return found->second;
}

std::int64_t price(const std::string& word) { return word == ABSENT ? 0 : std::stoll(word); }

int reference(const std::string& word) {
  if (word.empty() || word.front() != '#') {
    throw std::runtime_error("expected an order reference, got " + word);
  }
  return std::stoi(word.substr(1));
}

// Fixtures write the time in force short. Four names, so they are written out here.
TimeInForce::Value timeInForce(const std::string& word) {
  if (word == "GTC") return TimeInForce::GOOD_TILL_CANCEL;
  if (word == "DAY") return TimeInForce::DAY;
  if (word == "IOC") return TimeInForce::IMMEDIATE_OR_CANCEL;
  if (word == "FOK") return TimeInForce::FILL_OR_KILL;
  throw std::runtime_error("unknown time in force " + word);
}

Side::Value sideOf(const std::string& word) {
  for (const Side::Value value : {Side::BUY, Side::SELL}) {
    if (word == Side::c_str(value)) {
      return value;
    }
  }
  throw std::runtime_error("unknown side " + word);
}

PricingInstruction::Value pricingOf(const std::string& word) {
  for (const PricingInstruction::Value value :
       {PricingInstruction::LIMIT, PricingInstruction::MARKET}) {
    if (word == PricingInstruction::c_str(value)) {
      return value;
    }
  }
  throw std::runtime_error("unknown pricing instruction " + word);
}

SessionState::Value sessionStateOf(const std::string& word) {
  for (const SessionState::Value value :
       {SessionState::PRE_OPEN, SessionState::OPENING_AUCTION, SessionState::CONTINUOUS,
        SessionState::CLOSING_AUCTION, SessionState::HALTED, SessionState::CLOSED}) {
    if (word == SessionState::c_str(value)) {
      return value;
    }
  }
  throw std::runtime_error("unknown session state " + word);
}

AllocationAlgorithm::Value allocationOf(const std::string& word) {
  for (const AllocationAlgorithm::Value value :
       {AllocationAlgorithm::PRICE_TIME, AllocationAlgorithm::PRO_RATA}) {
    if (word == AllocationAlgorithm::c_str(value)) {
      return value;
    }
  }
  throw std::runtime_error("unknown allocation algorithm " + word);
}

bool contains(const std::vector<std::string>& words, const char* wanted) {
  for (const std::string& candidate : words) {
    if (candidate == wanted) {
      return true;
    }
  }
  return false;
}

std::size_t length(const std::uint64_t encodedLength) {
  return MessageHeader::encodedLength() + encodedLength;
}

}  // namespace

std::size_t CommandWriter::write(const Command& command) {
  sequence_++;
  switch (command.directive) {
    case Directive::INSTRUMENT:
      return instrument(command.arguments);
    case Directive::SESSION:
      return session(command.arguments);
    case Directive::NEW:
      return newOrder(command.arguments);
    case Directive::CANCEL:
      return cancel(command.arguments);
    case Directive::REPLACE:
      return replace(command.arguments);
    case Directive::MASSCANCEL:
      return massCancel(command.arguments);
  }
  throw std::runtime_error("unreachable");
}

std::size_t CommandWriter::instrument(const std::vector<std::string>& arguments) {
  const Options given = options(arguments, 0);
  protocol::InstrumentDefinition encoder;
  encoder.wrapAndApplyHeader(buffer_.data(), 0, buffer_.size());
  encoder.frame().instrumentId(INSTRUMENT_ID).sequence(sequence_);
  encoder.tickSize(number(given, "tick"))
      .lotSize(number(given, "lot"))
      .minPrice(number(given, "min"))
      .maxPrice(number(given, "max"))
      .priceScale(static_cast<std::uint8_t>(number(given, "scale")))
      .bandWidth(number(given, "band"))
      .openingReference(number(given, "open"))
      .allocation(allocationOf(word(given, "alloc")));
  return length(encoder.encodedLength());
}

std::size_t CommandWriter::session(const std::vector<std::string>& arguments) {
  protocol::SessionStateChange encoder;
  encoder.wrapAndApplyHeader(buffer_.data(), 0, buffer_.size());
  encoder.frame().instrumentId(INSTRUMENT_ID).sequence(sequence_);
  encoder.state(sessionStateOf(arguments.front()));
  return length(encoder.encodedLength());
}

std::size_t CommandWriter::newOrder(const std::vector<std::string>& arguments) {
  const Options given = options(arguments, 5);
  const int reference = ++orders_;
  const int participant = static_cast<int>(number(given, "p", DEFAULT_PARTICIPANT));
  references_.declare(reference, participant);

  protocol::NewOrder encoder;
  encoder.wrapAndApplyHeader(buffer_.data(), 0, buffer_.size());
  encoder.frame().instrumentId(INSTRUMENT_ID).sequence(sequence_);
  encoder.clientOrderId(reference)
      .participantId(participant)
      .side(sideOf(arguments[0]))
      .pricing(pricingOf(arguments[1]))
      .timeInForce(timeInForce(arguments[2]))
      .price(price(arguments[3]))
      .quantity(std::stoll(arguments[4]))
      .minQuantity(number(given, "min", 0))
      .displayQuantity(number(given, "display", 0))
      .triggerPrice(number(given, "trigger", 0))
      .smpId(number(given, "smp", 0));
  encoder.flags().clear().postOnly(contains(arguments, "POST_ONLY"));
  return length(encoder.encodedLength());
}

std::size_t CommandWriter::cancel(const std::vector<std::string>& arguments) {
  const int named = reference(arguments.front());
  const Options given = options(arguments, 1);
  protocol::CancelOrder encoder;
  encoder.wrapAndApplyHeader(buffer_.data(), 0, buffer_.size());
  encoder.frame().instrumentId(INSTRUMENT_ID).sequence(sequence_);
  encoder.clientOrderId(named).participantId(
      static_cast<std::uint32_t>(number(given, "p", references_.participant(named))));
  return length(encoder.encodedLength());
}

std::size_t CommandWriter::replace(const std::vector<std::string>& arguments) {
  const int named = reference(arguments.front());
  protocol::ReplaceOrder encoder;
  encoder.wrapAndApplyHeader(buffer_.data(), 0, buffer_.size());
  encoder.frame().instrumentId(INSTRUMENT_ID).sequence(sequence_);
  encoder.clientOrderId(named)
      .participantId(references_.participant(named))
      .quantity(std::stoll(arguments[1]))
      .price(price(arguments[2]));
  return length(encoder.encodedLength());
}

std::size_t CommandWriter::massCancel(const std::vector<std::string>& arguments) {
  const Options given = options(arguments, 0);
  protocol::MassCancel encoder;
  encoder.wrapAndApplyHeader(buffer_.data(), 0, buffer_.size());
  encoder.frame().instrumentId(INSTRUMENT_ID).sequence(sequence_);
  encoder.clientOrderId(0).participantId(static_cast<std::uint32_t>(number(given, "p")));
  return length(encoder.encodedLength());
}

}  // namespace io::github::giovanicaprison::matching::conformance

File: java/matching-conformance/src/main/java/io/github/giovanicaprison/matching/conformance/CommandWriter.java
package io.github.giovanicaprison.matching.conformance;

import io.github.giovanicaprison.matching.protocol.AllocationAlgorithm;
import io.github.giovanicaprison.matching.protocol.CancelOrderEncoder;
import io.github.giovanicaprison.matching.protocol.InstrumentDefinitionEncoder;
import io.github.giovanicaprison.matching.protocol.MassCancelEncoder;
import io.github.giovanicaprison.matching.protocol.MessageHeaderEncoder;
import io.github.giovanicaprison.matching.protocol.NewOrderEncoder;
import io.github.giovanicaprison.matching.protocol.PricingInstruction;
import io.github.giovanicaprison.matching.protocol.ReplaceOrderEncoder;
import io.github.giovanicaprison.matching.protocol.SessionState;
import io.github.giovanicaprison.matching.protocol.SessionStateChangeEncoder;
import io.github.giovanicaprison.matching.protocol.Side;
import io.github.giovanicaprison.matching.protocol.TimeInForce;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.agrona.ExpandableArrayBuffer;
import org.agrona.MutableDirectBuffer;

/**
 * Turns a fixture command into the bytes an engine receives.
 *
 * <p>An order reference is the client order id the harness gave that order, so a cancel or a
 * replace needs nothing an engine has to report first. That is what makes a fixture replayable
 * against any implementation without knowing how it numbers anything.
 *
 * <p>The instrument id and the input sequence are the harness's, not the fixture's. Nothing a
 * fixture writes decides them, which is the arrangement the engine always has: input arrives
 * already sequenced.
 *
 * <p>A refusal is rendered by the client order id, so {@code REJECTED #1 UNKNOWN_ORDER} names the
 * order the command meant rather than the position in the file it occupied.
 */
final class CommandWriter {

  private static final int INSTRUMENT_ID = 1;

  /** Fixtures write the time in force short. Four names, so they are written out here. */
  private static final Map<String, TimeInForce> TIME_IN_FORCE =
      Map.of(
          "GTC", TimeInForce.GOOD_TILL_CANCEL,
          "DAY", TimeInForce.DAY,
          "IOC", TimeInForce.IMMEDIATE_OR_CANCEL,
          "FOK", TimeInForce.FILL_OR_KILL);

  private static final String ABSENT = "-";
  private static final int DEFAULT_PARTICIPANT = 1;

  private final MutableDirectBuffer buffer = new ExpandableArrayBuffer(256);
  private final MessageHeaderEncoder header = new MessageHeaderEncoder();
  private final InstrumentDefinitionEncoder instrument = new InstrumentDefinitionEncoder();
  private final SessionStateChangeEncoder session = new SessionStateChangeEncoder();
  private final NewOrderEncoder newOrder = new NewOrderEncoder();
  private final CancelOrderEncoder cancel = new CancelOrderEncoder();
  private final ReplaceOrderEncoder replace = new ReplaceOrderEncoder();
  private final MassCancelEncoder massCancel = new MassCancelEncoder();

  private final References references;
  private long sequence;
  private int orders;

  CommandWriter(final References references) {
    this.references = references;
  }

  MutableDirectBuffer buffer() {
    return buffer;
  }

  /** Encodes one command at offset zero and returns its length in bytes. */
  int write(final Fixture.Command command) {
    sequence++;
    return switch (command.directive()) {
      case INSTRUMENT -> instrument(command.arguments());
      case SESSION -> session(command.arguments());
      case NEW -> newOrder(command.arguments());
      case CANCEL -> cancel(command.arguments());
      case REPLACE -> replace(command.arguments());
      case MASSCANCEL -> massCancel(command.arguments());
    };
  }

  private int instrument(final List<String> arguments) {
    final Map<String, String> options = options(arguments);
    instrument.wrapAndApplyHeader(buffer, 0, header);
    instrument.frame().instrumentId(INSTRUMENT_ID).sequence(sequence);
    instrument
        .tickSize(number(options, "tick"))
        .lotSize(number(options, "lot"))
        .minPrice(number(options, "min"))
        .maxPrice(number(options, "max"))
        .priceScale((short) number(options, "scale"))
        .bandWidth(number(options, "band"))
        .openingReference(number(options, "open"))
        .allocation(AllocationAlgorithm.valueOf(word(options, "alloc")));
    return length(instrument.encodedLength());
  }

  private int session(final List<String> arguments) {
    session.wrapAndApplyHeader(buffer, 0, header);
    session.frame().instrumentId(INSTRUMENT_ID).sequence(sequence);
    session.state(SessionState.valueOf(arguments.getFirst()));
    return length(session.encodedLength());
  }

  private int newOrder(final List<String> arguments) {
    final Map<String, String> options = options(arguments.subList(5, arguments.size()));
    final int reference = ++orders;
    final int participant = (int) number(options, "p", DEFAULT_PARTICIPANT);
    references.declare(reference, participant);

    newOrder.wrapAndApplyHeader(buffer, 0, header);
    newOrder.frame().instrumentId(INSTRUMENT_ID).sequence(sequence);
    newOrder
        .clientOrderId(reference)
        .participantId(participant)
        .side(Side.valueOf(arguments.get(0)))
        .pricing(PricingInstruction.valueOf(arguments.get(1)))
        .timeInForce(timeInForce(arguments.get(2)))
        .price(price(arguments.get(3)))
        .quantity(Long.parseLong(arguments.get(4)))
        .minQuantity(number(options, "min", 0))
        .displayQuantity(number(options, "display", 0))
        .triggerPrice(number(options, "trigger", 0))
        .smpId(number(options, "smp", 0));
    newOrder.flags().clear().postOnly(arguments.contains("POST_ONLY"));
    return length(newOrder.encodedLength());
  }

  private int cancel(final List<String> arguments) {
    final int reference = reference(arguments.getFirst());
    final Map<String, String> options = options(arguments.subList(1, arguments.size()));
    cancel.wrapAndApplyHeader(buffer, 0, header);
    cancel.frame().instrumentId(INSTRUMENT_ID).sequence(sequence);
    cancel
        .clientOrderId(reference)
        .participantId(number(options, "p", references.participant(reference)));
    return length(cancel.encodedLength());
  }

  private int replace(final List<String> arguments) {
    final int reference = reference(arguments.getFirst());
    replace.wrapAndApplyHeader(buffer, 0, header);
    replace.frame().instrumentId(INSTRUMENT_ID).sequence(sequence);
    replace
        .clientOrderId(reference)
        .participantId(references.participant(reference))
        .quantity(Long.parseLong(arguments.get(1)))
        .price(price(arguments.get(2)));
    return length(replace.encodedLength());
  }

  private int massCancel(final List<String> arguments) {
    final Map<String, String> options = options(arguments);
    massCancel.wrapAndApplyHeader(buffer, 0, header);
    massCancel.frame().instrumentId(INSTRUMENT_ID).sequence(sequence);
    massCancel.clientOrderId(0).participantId(number(options, "p"));
    return length(massCancel.encodedLength());
  }

  private static TimeInForce timeInForce(final String word) {
    final TimeInForce timeInForce = TIME_IN_FORCE.get(word);
    if (timeInForce == null) {
      throw new IllegalArgumentException("unknown time in force " + word);
    }
    return timeInForce;
  }

  private static long price(final String word) {
    return ABSENT.equals(word) ? 0 : Long.parseLong(word);
  }

  private static int reference(final String word) {
    if (!word.startsWith("#")) {
      throw new IllegalArgumentException("expected an order reference, got " + word);
    }
    return Integer.parseInt(word.substring(1));
  }

  private static Map<String, String> options(final List<String> words) {
    final Map<String, String> options = new HashMap<>();
    for (final String word : words) {
      final int equals = word.indexOf('=');
      if (equals > 0) {
        options.put(word.substring(0, equals), word.substring(equals + 1));
      }
    }
    return options;
  }

  private static long number(final Map<String, String> options, final String key) {
    final String value = options.get(key);
    if (value == null) {
      throw new IllegalArgumentException("missing " + key);
    }
    return Long.parseLong(value);
  }

  private static long number(
      final Map<String, String> options, final String key, final long fallback) {
    final String value = options.get(key);
    return value == null ? fallback : Long.parseLong(value);
  }

  private static String word(final Map<String, String> options, final String key) {
    final String value = options.get(key);
    if (value == null) {
      throw new IllegalArgumentException("missing " + key);
    }
    return value;
  }

  private int length(final int encodedLength) {
    return MessageHeaderEncoder.ENCODED_LENGTH + encodedLength;
  }
}
